File: app/src/main/java/com/irishinterest/irishinterest/fragments/authors/helper/AuthorsListAdapter.java
package com.irishinterest.irishinterest.fragments.authors.helper;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.irishinterest.irishinterest.R;
import com.irishinterest.irishinterest.BooksActivity;
import com.irishinterest.irishinterest.model.authors.Author;


public class AuthorsListAdapter extends PagedListAdapter<Author, AuthorsListAdapter.ViewHolder> {
    private Context context;

    public AuthorsListAdapter(Context context) {
        super(DIFF_CALLBACK);
        this.context = context;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public ViewHolder(View v) {
            super(v);
        }

        private TextView authorName;

        public void bindView(Author author){
            authorName = itemView.findViewById(R.id.authorName);
            authorName.setText(author.getFirstname() + " " + author.getLastname());
            authorName.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context, BooksActivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    intent.putExtra("type", "authors");
                    if(author.getFirstname() != null) {
                        intent.putExtra("firstName", author.getFirstname().replace("'", ""));
                    } else {
                        intent.putExtra("firstName", "");
                    }
                    if(author.getFirstname() != null){
                        intent.putExtra("lastName", author.getLastname().replace("'", ""));
                    } else {
                        intent.putExtra("lastName", "");
                    }

                    intent.putExtra("id", String.valueOf(author.getId()));
                    intent.putExtra("profile", author.getProfile() == null || author.getProfile().isEmpty()? "Author information pending." : author.getProfile());
                    intent.putExtra("authorImage", author.getImage() == null ? null : author.getImage().toString());
                    context.startActivity(intent);
                }
            });
        }
    }

    @NonNull
    @Override
    public AuthorsListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.author_list_item, parent, false);
        return new AuthorsListAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull AuthorsListAdapter.ViewHolder holder, int position) {
        Author author = getItem(position);
        if(author != null) {
            holder.bindView(author);
        }
    }


    private static DiffUtil.ItemCallback<Author> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<Author>() {
                @Override
                public boolean areItemsTheSame(Author oldItem, Author newItem) {
                    return oldItem.getId() == newItem.getId();
                }

                @RequiresApi(api = Build.VERSION_CODES.KITKAT)
                @Override
                public boolean areContentsTheSame(Author oldItem, Author newItem) {
                    return oldItem.equals(newItem);
                }
            };

}

File: app/src/main/java/com/irishinterest/irishinterest/fragments/authors/helper/AuthorsListAdapter.kt
package com.irishinterest.irishinterest.fragments.authors.helper

import android.content.Context
import androidx.paging.PagedListAdapter
import com.irishinterest.irishinterest.model.authors.Author
import com.irishinterest.irishinterest.fragments.authors.helper.AuthorsListAdapter
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import com.irishinterest.irishinterest.R
import android.content.Intent
import com.irishinterest.irishinterest.BooksActivity
import android.view.ViewGroup
import android.view.LayoutInflater
import androidx.recyclerview.widget.DiffUtil
import androidx.annotation.RequiresApi
import android.os.Build
import android.view.View

class AuthorsListAdapter(private val context: Context) :
    PagedListAdapter<Author, AuthorsListAdapter.ViewHolder>(
        DIFF_CALLBACK
    ) {
    inner class ViewHolder(v: View?) : RecyclerView.ViewHolder(v!!) {
        fun bindView(author: Author) {
            val authorName: TextView = itemView.findViewById(R.id.authorName)
            authorName.setText(author.firstname + " " + author.lastname)
            authorName.setOnClickListener(View.OnClickListener {
                val intent = Intent(context, BooksActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                intent.putExtra("type", "authors")
                if (author.firstname != null) {
                    intent.putExtra("firstName", author.firstname!!.replace("'", ""))
                } else {
                    intent.putExtra("firstName", "")
                }
                if (author.firstname != null) {
                    intent.putExtra("lastName", author.lastname!!.replace("'", ""))
                } else {
                    intent.putExtra("lastName", "")
                }
                intent.putExtra("id", author.id.toString())
                intent.putExtra(
                    "profile",
                    if (author.profile == null || author.profile!!.isEmpty()) "Author information pending." else author.profile
                )
                intent.putExtra(
                    "authorImage",
                    if (author.image == null) null else author.image.toString()
                )
                context.startActivity(intent)
            })
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.author_list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val author = getItem(position)
        if (author != null) {
            holder.bindView(author)
        }
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<Author> =
            object : DiffUtil.ItemCallback<Author>() {
                override fun areItemsTheSame(oldItem: Author, newItem: Author): Boolean {
                    return oldItem.id === newItem.id
                }

                @RequiresApi(api = Build.VERSION_CODES.KITKAT)
                override fun areContentsTheSame(oldItem: Author, newItem: Author): Boolean {
                    return oldItem.equals(newItem)
                }
            }
    }
}